File: src/main/go/leetcode/lists/topinterview/LeetCode340.go
package topinterview

func lengthOfLongestSubstringKDistinct(s string, k int) int {
	if len(s) == 0 || k == 0 {
		return 0
	}
	left, right, n := 0, 0, len(s)
	mp := make(map[byte]int)
	res := 0
	charCount := 0
	for right < n {
		mp[s[right]]++
		if mp[s[right]] == 1 {
			charCount++
		}
		for charCount > k {
			if mp[s[left]] == 1 {
				charCount--
			}
			mp[s[left]]--
			left++
		}
		if charCount <= k {
			res = Max(right-left+1, res)
		}
		right++
	}
	return res
}

File: src/main/java/leetcode/lists/topinterview/LeetCode340.java
package leetcode.lists.topinterview;

import java.util.HashMap;
import java.util.Map;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/6 18:31
 */
public class LeetCode340 {

    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        Map<Character, Integer> counts = new HashMap<>();
        int maxLength = 0;
        int left = 0;
        int right = 0;
        while (right < s.length()) {
            char curr = s.charAt(right);
            counts.put(curr, counts.getOrDefault(curr, 0) + 1);
            while (counts.size() > k) {
                char prev = s.charAt(left);
                counts.put(prev, counts.get(prev) - 1);
                if (counts.get(prev) == 0) {
                    counts.remove(prev);
                }
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
            right++;
        }
        return maxLength;
    }
}
